Tracker/Modules/MainScreenModule/TrackerModule/TrackerMainScreen/TrackersViewModel.swift

```diff
@@ -0,0 +1,153 @@
+//
+//  TrackersViewModel.swift
+//  Tracker
+//
+//  Created by Aleksandr Eliseev on 09.05.2023.
+//
+
+import UIKit
+
+protocol TrackerMainScreenDelegate: AnyObject {
+    func saveTracker(tracker: Tracker, to categoryName: String)
+}
+
+final class TrackersViewModel: NSObject {
+            
+    @Observable
+    private (set) var completedTrackers: Set<TrackerRecord> = []
+    @Observable
+    private (set) var visibleCategories: [TrackerCategory] = [] {
+        didSet {
+            checkForEmptyState()
+        }
+    }
+    @Observable
+    private (set) var emptyStackViewIsHidden: Bool = false
+        
+    override init() {
+        super.init()
+        self.fetchTrackers()
+        self.fetchRecords()
+    }
+    
+    private lazy var coreDataManager: CoreDataManagerProtocol = {
+        coreDataManager = CoreDataManager(coreDataManagerDelegate: self)
+        return coreDataManager
+    }()
+    
+    func getCurrentDate(from date: Date) -> Date? {
+        return date.customlyFormatted()
+    }
+    
+    func filterTrackers(with string: String) {
+        var temporaryCategories: [TrackerCategory] = []
+        
+        for category in visibleCategories {
+            let filteredTrackers = category.trackers.filter({ $0.name.lowercased().contains(string.lowercased()) })
+            if !filteredTrackers.isEmpty {
+                let filteredCategory = TrackerCategory(name: category.name, trackers: filteredTrackers)
+                temporaryCategories.append(filteredCategory)
+                visibleCategories = temporaryCategories
+            } else {
+                visibleCategories = temporaryCategories
+            }
+        }
+    }
+    
+    private func fetchTrackers() {
+        visibleCategories = coreDataManager.fetchedCategories
+    }
+    
+    private func fetchRecords() {
+        completedTrackers = coreDataManager.fetchedRecords
+    }
+}
+
+// MARK: - Ext TrackerMainScreenDelegate
+extension TrackersViewModel: TrackerMainScreenDelegate {
+    func saveTracker(tracker: Tracker, to categoryName: String) {
+        DispatchQueue.main.async { [weak self] in
+            guard let self = self else { return }
+            do {
+                try self.coreDataManager.saveTracker(tracker: tracker, to: categoryName)
+                self.checkForScheduledTrackers(with: getCurrentDate(from: Date()))
+            } catch {
+                print("saveTracker failed")
+            }
+        }
+        
+    }
+}
+
+// MARK: - Ext Collection view checkers
+extension TrackersViewModel {
+    func checkForEmptyState(){
```

Comment:
Отступы вокруг любой фигурной строки должны быть
```suggestion
    func checkForEmptyState() {
```